marlowe-playground-client/src/Simulation.purs

```diff
@@ -253,6 +278,79 @@ handleAction settings AnalyseReachabilityContract = do
 
 handleAction _ Save = pure unit
 
+setOraclePrice ::
+  forall m.
+  MonadAff m =>
+  SPSettings_ SPParams_ -> HalogenM State Action ChildSlots Void m Unit
+setOraclePrice settings = do
+  exState <- use (_currentMarloweState <<< _executionState)
+  case exState of
+    Just es -> do
+      let
+        (Parties actions) = es.possibleActions
+      case Map.lookup (Role "kraken") actions of
+        Just acts -> do
+          case Array.head (Map.toUnfoldable acts) of
+            Just (Tuple (ChoiceInputId choiceId@(ChoiceId pair _) bounds) _) -> do
+              price <- getPrice settings "kraken" pair
+              liftEffect $ log ("setting choice " <> show choiceId <> " with price " <> show price)
+              handleAction settings (SetChoice choiceId price)
+            _ -> pure unit
+        Nothing -> pure unit
+    Nothing -> pure unit
```

Comment:
In the purescript codebase, we've gone down the route of lenses so lets use them. In this case you only ever act in one of the branches, in this case there are plenty of nice lens combinators that can get you to this point and not need to do loads of pattern matching. This is a guess but something like:
```purescript
mChoice <- peruse (_currentMarloweState <<< _executionState <<< _Just <<< _possibleActions <<< ix (Role "kraken") <<< to Map.keys <<< _Head)
for_ mChoice \(ChoiceInputId choiceId@(ChoiceId pair _) bounds) -> do
              price <- getPrice settings "kraken" pair
              liftEffect $ log ("setting choice " <> show choiceId <> " with price " <> show price)
              handleAction settings (SetChoice choiceId price)
```

Reply:
After some thinking I suggest to leave it like this. I'll add more exchanges later, so I'll need to check agains a set of values. I guess it'd be more "understandable" in this naive implementation rather than using advanced profunctor optics.

Reply:
The issue is that it is perhaps more understandable to readers who aren't used to optics but people who are contributing to this particular code base either are used to them or will be soon since the code base is heavily reliant on them.

Reply:
this is just for discussion since we are going to merge this as is but what might be nicer is to `derive newtype Index Parties`, then it could be used in a prism in a very nice way.